x/exchange/types/matching.go

```diff
@@ -29,62 +119,35 @@ func NewMatchingContext(market Market, halveFees bool) *MatchingContext {
 	}
 }
 
-func (ctx *MatchingContext) feeRate(isMaker bool) sdk.Dec {
-	if isMaker {
-		return ctx.makerFeeRate
-	}
-	return ctx.takerFeeRate
-}
-
-func (ctx *MatchingContext) FillOrder(order *MemOrder, qty, price sdk.Dec, isMaker bool) {
-	executableQty := order.ExecutableQuantity()
-	if qty.GT(executableQty) { // sanity check
-		panic("open quantity is less than quantity")
-	}
-	if order.isMaker != nil && isMaker != *order.isMaker { // sanity check
-		panic("an order's isMaker must be consistent under one matching context")
-	}
-	_, pays, receives, fee := ctx.fillOrder(order.typ, order.isBuy, qty, price, isMaker)
-	order.paid = order.paid.Add(pays)
-	order.remainingDeposit = order.remainingDeposit.Sub(pays)
-	order.received = order.received.Add(receives)
-	order.fee = order.fee.Add(fee)
-	order.executedQty = order.executedQty.Add(qty)
-	order.isMatched = true
-	order.isMaker = &isMaker
-}
-
-func (ctx *MatchingContext) fillOrder(orderType MemOrderType, isBuy bool, qty, price sdk.Dec, isMaker bool) (executedQuote, pays, receives, fee sdk.Dec) {
-	executedQuote = QuoteAmount(isBuy, price, qty)
-	if isBuy {
-		pays = executedQuote
-		receives = qty
-	} else {
-		pays = qty
-		receives = executedQuote
-	}
-	var feeRate sdk.Dec
+func (ctx *MatchingContext) FeeRate(orderType MemOrderType, isMaker bool) (feeRate sdk.Dec) {
 	if orderType == UserMemOrder {
-		feeRate = ctx.feeRate(isMaker)
-	} else {
 		if isMaker {
-			feeRate = ctx.takerFeeRate.Mul(ctx.orderSourceFeeRatio).Neg()
+			feeRate = ctx.makerFeeRate
+		} else {
+			feeRate = ctx.takerFeeRate
+		}
+	} else if orderType == OrderSourceMemOrder { // order source order
+		if isMaker {
+			feeRate = ctx.takerFeeRate.Neg().Mul(ctx.orderSourceFeeRatio)
 		} else {
 			feeRate = utils.ZeroDec
 		}
+	} else { // sanity check
+		panic("invalid order type")
 	}
-	if feeRate.IsNegative() {
-		fee = feeRate.MulTruncate(pays)
-		pays = pays.Add(fee)
-	} else if feeRate.IsPositive() {
-		receives, fee = DeductFee(receives, feeRate)
-	} else {
-		fee = utils.ZeroDec
+	return feeRate
+}
+
+func (ctx *MatchingContext) FillOrder(order *MemOrder, qty sdk.Int, price sdk.Dec, isMaker bool) {
+	executableQty := order.ExecutableQuantity()
```

Comment:
For sanity check, can we have checks to see if order.Price >= price for buy order and order.Price <= price for sell order?

Reply:
It's a good idea. I think the overhead will be small enough for those sanity checks.